transactions/btc.ts

```diff
@@ -526,6 +527,201 @@ export async function getBtcFeesForNonOrdinalBtcSend(
   }
 }
 
+function getTransactionMetadataForUtxos(
+  recipients: Recipient[],
+  selectedUtxos: UTXO[],
+  changeAddress: string,
+  feeRate: number,
+): TransactionUtxoSelectionMetadata | undefined {
+  const dummyPrivateKey = '0000000000000000000000000000000000000000000000000000000000000001';
+
+  const inputSum = selectedUtxos.reduce<BigNumber>((sum, utxo) => sum.plus(utxo.value), new BigNumber(0));
+  const recipientTotal = recipients.reduce<BigNumber>(
+    (sum, recipient) => sum.plus(recipient.amountSats),
+    new BigNumber(0),
+  );
+
+  // these are conservative estimates
+  const ESTIMATED_VBYTES_PER_OUTPUT = 45; // actually around 50
+  const ESTIMATED_VBYTES_PER_INPUT = 85; // actually around 89 or 90
+  const BITCOIN_DUST_VALUE = 1000;
+
+  const estimatedFees = new BigNumber(feeRate).times(
+    ESTIMATED_VBYTES_PER_OUTPUT * recipients.length + ESTIMATED_VBYTES_PER_INPUT * selectedUtxos.length,
+  );
+  // ensure that the UTXOs can cover the expected outputs
+  if (!inputSum.gt(recipientTotal.plus(estimatedFees))) return undefined;
+
+  {
+    // try transaction with change
+    const tx = createTransaction(dummyPrivateKey, selectedUtxos, recipientTotal, recipients, changeAddress, 'Mainnet');
+
+    tx.sign(hex.decode(dummyPrivateKey));
+    tx.finalize();
+
+    const txSize = tx.vsize;
+    let sentSats = new BigNumber(0);
+
+    for (let i = 0; i < tx.outputsLength; i++) {
+      const output = tx.getOutput(i);
+      sentSats = sentSats.plus(new BigNumber((output.amount ?? 0n).toString()));
+    }
+
+    const change = sentSats.minus(recipientTotal);
+    const fee = new BigNumber(txSize).times(feeRate);
+
+    // check if there is change and if the change is greater than the vsize*feeRate + dust rate
+    if (change.gt(fee.plus(BITCOIN_DUST_VALUE))) {
+      const changeWithoutFee = change.minus(fee);
+      return {
+        selectedUtxos,
+        fee: fee.toNumber(),
+        feeRate: feeRate,
+        change: changeWithoutFee.toNumber(),
+      };
+    }
+  }
+
+  {
+    // try txn without change
```

Comment:
not sure I follow the logic compared to PR description:
> This means that the following are the desired outcomes (1 being more favourable), while keeping the fee at a minimum:
> 1. Having one or more input UTXOs which match the recipient sum+fees, as close as possible. This may result in a slightly higher fee rate than desired but would eliminate the change output (decreasing fees even though having a higher fee rate) and avoiding dust UTXOs.
> 2. Having the least number of inputs resulting in the biggest change output. This would use up UTXOs efficiently, while also ensuring that future transactions have a new UTXO with the highest value possible for more efficient future spend.

isn't this function trying 2. (with change) and then 1. (no change) ?

Reply:
Yeah, this function is doing that. It will first ensure that the transaction cannot be made with change before trying with no change. The second part would always return a valid transaction if the first one would return one, except the bottom one would result in a potentially very large fee in that case.

So, all this function does is check if we can use the UTXOs for the transaction, and whether there would be change; it doesn't actually apply the logic in the PR description. That logic is applied in the `selectOptimalUtxos` method in this section:
```
/**
     * we want to:
     * - minimise fees
     * - have zero change or maximise change
     * - while staying above the selected fee rate
     **/
    if (!bestSelectionData) {
      bestSelectionData = nextSelectionData;
    } else if (bestSelectionData.fee > nextSelectionData.fee) {
      bestSelectionData = nextSelectionData;
    } else if (nextSelectionData.fee === bestSelectionData.fee) {
      if (bestSelectionData.change < nextSelectionData.change) {
        bestSelectionData = nextSelectionData;
      }
    } else {
      return bestSelectionData;
    }
```

A transaction without change would have a lower fee as it has one fewer outputs.

Reply:
Let me know if there are any comments I can put in the code that would clear things up for future devs 🙏